packages/jest-phabricator-coverage/example/JestUnitTestEngine.php

```diff
@@ -0,0 +1,173 @@
+<?php
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+/**
+ * JestUnitTestEngine
+ */
+class JestUnitTestEngine extends ArcanistBaseUnitTestEngine {
+  const PROCESSOR = 'jest/packages/jest-phabricator-coverage/index.js';
+  const JEST_PATH = 'jest/packages/jest/bin/jest.js';
+
+  private function getRoot() {
+    return $this->getWorkingCopy()->getProjectRoot();
+  }
+
+  private function getOutputJSON() {
+    return $this->getRoot() . '/output.json';
+  }
+
+  private function getFutureResults($future) {
+    list($stdout, $stderr) = $future->resolvex();
+    $output_JSON = $this->getOutputJSON();
+    $report_path_exists = file_exists($output_JSON);
+    $raw_results = null;
+
+    if ($report_path_exists) {
+      $raw_results = json_decode(
+        Filesystem::readFile($output_JSON),
+        true
+      )['phabricatorReport'];
+      Filesystem::remove($output_JSON);
+    } else {
+      $raw_results = json_decode($stdout, true);
+    }
+
+    if (!is_array($raw_results)) {
+      throw new Exception("Unit test script emitted invalid JSON: {$stdout}");
+    }
+
+    $results = array();
+    foreach ($raw_results as $result) {
+      $test_result = new ArcanistUnitTestResult();
+      $test_result->setName($result['name']);
+      $succeed = isset($result['status']) && $result['status'] == 'passed';
+      $test_result->setResult(
+        $succeed ?
+        ArcanistUnitTestResult::RESULT_PASS :
+        ArcanistUnitTestResult::RESULT_FAIL
+      );
+
+      if (isset($result['coverage'])) {
+        $coverage = array();
+        $root = $this->getRoot() . '/';
+        foreach ($result['coverage'] as $file_path => $coverage_data) {
+          if (substr($file_path, 0, strlen($root)) == $root) {
+            $file_path = substr($file_path, strlen($root));
+          }
+          $coverage[$file_path] = $coverage_data;
+       }
+       $test_result->setCoverage($coverage);
+     }
+     $test_result->setUserData($result['message']);
+     $results[] = $test_result;
+   }
+
+    return $results;
+  }
+
+  private function runCommands($commands) {
+    $futures = array();
+    foreach ($commands as $command) {
+      $bin = $command['bin'];
+      $options = implode(' ', $command['options']);
+      $paths = $command['paths'];
+      $futures[] = new ExecFuture("{$bin} {$options} %Ls", $paths);
+    }
+
+    $this->console = PhutilConsole::getConsole();
+
+    // Pass stderr through so we can give the user updates on test
+    // status as tests run.
+    $completed = array();
+    foreach (Futures($futures)->setUpdateInterval(0.2) as $_) {
```

Comment:
I think `Futures` is deprecated so it should be changed to  `new FutureIterator($futures)` 

See: https://secure.phabricator.com/D11081